### ISSUE TITLE
Building automatic unittests on controller package

### ISSUE THREAD

**Issue by mdantonio** [rapydo/issues#38]
I'm trying to implement a set of unittests on controller, but i found several problems

First of all: a refactor of arguments management was required to be able to create Application objects by passing a list of parameters (and emulate CLI execution)

Now we are able to do this:

```
    command = "rapydo init".split(" ")
    arguments = ArgParser(args=command)

    try:
        Application(arguments)
    # NOTE: docker-compose calls SystemExit at the end of the command...
    except SystemExit:
        log.info('completed')
```

After the execution of a command we can verify it in two ways:
- verify if the actions are executed on the system (docker ps to verify if containers are running, check if folder exists, etc)
- checks on command output (logs)

The first way is stronger but also more complex. For now i'm experimenting with the second way

**First problem**: by default pytest captures both stdout and sterr (and print them only in case of failure)

Solutions: stdout and sterr can be accessed by using the capsys and capfd objects [pytest docs for details](https://docs.pytest.org/en/latest/capture.html). With basic configuration i'm not able to extract stderr from capfd (always empty...) so I added the -s flags (disable all capturing). In this ways capfd started to work... but this solution is absolutely counter-intuitive (I have to disable capturing to get captured logs??). Since @pdonorio already introduced the -s flag to solve the ` 'io.UnsupportedOperation: redirected stdin is pseudofile, has no fileno()' on docker compose` error we are ok with this flag... However i don't really understand this mechnism

**Second problem**: the captured output is based on our logging format (with date, colors and many and other awkward elements)

Solutions: i tried to intercept all loggers before they start to print to change the format of their handler... But after some tries, I failed and give up. I applied a very old style workaround: loop against the error list to cut the message just after the ] character to obtain a cleaner message. This solution works... but is really really ugly

**Third problem (unsolved for now)**: the creation of a container (for example the bower container during the init process) stops the logging (i.e. Captured logs stop just after the container). In this way if bower is executed the output log does not contain the last message: "Project initialized"

**Comment by pdonorio** [comment]
> the captured output is based on our logging format

We may avoid that format when using unittests. Use [this variable](https://github.com/rapydo/do/blob/master/controller/__init__.py#L13-L16) to decide.

> the creation of a container stops the logging

probably we must silence or redirect docker logging while testing

**Comment by mdantonio** [comment]
>> the creation of a container stops the logging

> probably we must silence or redirect docker logging while testing

I found and fixed the problem: docker-compose run exits after the execution, so the rapydo-controller stops its execution. Added a try-except to ignore the SystemExit and now it works

> We may avoid that format when using unittests. Use this variable to decide.

Such variable is at controller level, we need it in utils/log 

And since we set the format (ad other confs) via logging.ini, which solution could be the best one? use a logging_test.ini file ?

**Comment by pdonorio** [comment]
> use a logging_test.ini file

Yes. But how do you specify that in `utilities.logs`?

**Comment by mdantonio** [comment]
I don't know... that's why now i leave untouched the log format and then I strip out from the log output everything expect the message... It is a bad approach, but i don't know how to hack from outside the logging, to force a different format

**Comment by mdantonio** [comment]
Could we adopt an env variable to communicate with utils/logs?

**Comment by pdonorio** [comment]
We already avoid colors in nose2 unittests:

https://github.com/rapydo/build-templates/blob/master/backend/tests.sh#L6-L7

**Comment by mdantonio** [comment]
Introduced an env variable (TESTING) to simplify the logging output for testing purpose

At the moment the controller tests fail because travis uses the last released package (0.5.1) and not the current utils version (with the above implementation)

Since a number of tests have been implementated for controller, we can close here.
